### DISCUSSION TITLE
`pm` can be run in termux without root/adb

### DISCUSSION THREAD

**Discussion by Efreak** [termux/termux-packages#8292] [Q&A] [answered]
`pm` does not run in termux without root/adb access. This has been an issue for quite a long time; the most recent issue is #8218.

I was surprised today to note that the [macchina](https://github.com/Macchina-CLI/macchina) is able to get the count of Android apps installed from within Termux. On looking at the source code, there doesn't seem to be anything strange going on, [it simply runs `pm list packages`](https://github.com/Macchina-CLI/libmacchina/blob/98540e4ea3f258583b4587ad72220f603d5de5cc/src/android/mod.rs#L400). Neither libmacchina nor macchina is doing anything weird; looking at `strace -fo macchina.trace` output, it's looking for a pm binary in the path, and eventually running `/data/data/com.termux/files/usr/bin/pm`.

I tried running /system/bin/sh with the same environment that libmacchina uses (by modifying it to run `env` and print the output), but it's still not working. I'm wondering if termux is doing something that is affecting the process that causes it to not run properly, and it works with macchina being statically compiled (I have no idea if this even makes any sense). While I do have shizuku installed and it allows termux, I don't have it running right now (adb is currently disabled as well).

Regardless, macchina shows that the `pm` command can be run successfully from inside Termux, without root or adb.

**Comment by agnostic-apollo** [comment, marked as answer]
This is because of selinux restrictions.

>Isolate untrusted app ptys from other domains.

> Add a create_pty() macro that allows a domain to
create and use its own ptys, isolated from the ptys
of any other domain, and use that macro for untrusted_app.
This permits the use of a pty by apps without opening up access
to ptys created by any other domain on the system.

https://cs.android.com/android/_/android/platform/system/sepolicy/+/2dc4acf33b78284f514fe9a6c5102cc783c4309f

https://cs.android.com/android/_/android/platform/system/sepolicy/+/07c3a5a5222c8c0306cd62575f7f9279f7ca6093

https://cs.android.com/android/platform/superproject/+/android-13.0.0_r31:system/sepolicy/private/untrusted_app_all.te;l=160

https://cs.android.com/android/platform/superproject/+/android-13.0.0_r31:system/sepolicy/public/shell.te;l=53

Basically, what this means is that user apps are assigned one of the `untrusted_app*` domains (source context), which shares policies with the `untrusted_app_all` domain and the `pty` devices created by them are assigned the `untrusted_app_all_devpts` target context. So only `untrusted_app*` can access the `untrusted_app_all_devpts` `pty` devices and not any other source domains. However, when `cmd`  command is run, there would be a source transition to `system_server` domain, which then prevents it from reading and writing to the `pty` device originally created by the `untrusted_app*` domains.

```shell
# Remove audit log rate limiting and then run command to check
$ adb shell "su -c 'auditctl -r 0'"
# Check denials
$ adb logcat -d | grep "avc: denied"
avc: denied { read write } for path="/dev/pts/0" dev="devpts" ino=3 scontext=u:r:system_server:s0 tcontext=u:object_r:untrusted_app_all_devpts:s0:c161,c256,c512,c768 tclass=chr_file permissive=0 app=com.termux
```

Known it for a while that you can bypass this restriction to run "some" `pm`/`cmd` commands that are not restricted by other selinux policies or android permissions by running them in a subshell by redirecting `stderr` to `stdout` (`2>&1`) and redirecting `stdin` to `/dev/null` (`</dev/null`). **Both** are necessary to prevent `read`/`write` of `chr_file`.

The `/system/bin/am` `starts-service` and `broadcast` commands can also be run this way, but not the `start-activity` command, since there is a calling `uid` comparison against the `com.android.shell` package, check https://github.com/termux/TermuxAm/issues/9#issuecomment-1624590083.

```shell
$ out="$(pm list packages --user 0 2>&1 </dev/null)"
$ echo "$out"
package:android
package:com.termux
...
$ out="$(pm path com.termux --user 0 2>&1 </dev/null)"
$ echo "$out"
package:/data/app/~~zAi6GFqEAPTkudspFt7usg==/com.termux-XCCktU51me6pb27dF1b82A==/base.apk
```

Note that for `pm list packages`, package visibility changes will apply if app is using `targetSdkVersion` `>= 30` on Android `>= 11`, unless `QUERY_ALL_PACKAGES` permissions is requested or packages are listed in `queries` element of `AndroidManifest.xml`.

https://github.com/termux/termux-app/wiki/Termux-Libraries#target-sdk-30-package-visibility

You can patch the `sepolicy` with root to remove this restriction by running `supolicy --live "allow system_server untrusted_app_all_devpts chr_file { read write }"`
